### PR TITLE
Fix restart

### PR DESCRIPTION
**Pull request by sam-github**
/to piscisaureus, do you have time to review, or should I punt to someone else?

connected to strongloop-internal/scrum-nodeops#858

### CODE REVIEW

**Comment by sam-github** [comment]
/to @rmg you modified one of our repos to read mocha output into tap, recall which it was?

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by rmg** [comment]
For mocha + tap:
 1. upgrade tap to tap@1
   * this is normally the hardest part, ironically
 2. create `test/test-mocha-tests.js`
 3. change scripts.test to `tap test/test-*.js`

test/test-mocha.js would look something like this:
```js
var path = require('path');
var tap = require('tap');

// mocha tests assume they are run from package root, not test/
var cwd = path.resolve(__dirname, '../');
var mocha = require.resolve('mocha/bin/_mocha');
var args = [
  mocha, '--reporter', 'tap',
];

tap.spawn(process.execPath, args.concat(['test/master.js']), {cwd: cwd}, test);
tap.spawn(process.execPath, args.concat(['test/require.js']), {cwd: cwd}, test);
```

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@rmg PTAL, tests should pass next round

File: lib/master.js [hunk 1]
```diff
@@ -23,7 +23,9 @@ function liveWorkerIds() {
       worker.process.exitCode == null;
     return isAlive;
   });
-  return workerIds;
+  return workerIds.sort(function(a, b) {
```
Comment (rmg) [inline lib/master.js]: Go home, JavaScript, you're drunk...

Fortunately, lodash's sort does the right thing:
```js
return _.sortBy(workerIds);
```

I'm not sure which is better, though... using Array.prototype.sort requires a numeric comparator, which one could naively assume isn't necessary, resulting in the reader assuming this is really a `.reverse()` done the "hard" way.

Using `lodash.sortyBy` without specifying a field to sort by raises the question of why `Array.prototype.sort` isn't being used..

Comment (rmg) [inline lib/master.js]: I guess what I'm suggesting is to add a comment complaining about the behaviour of `Array.prototype.sort` and how it sorts numbers as strings and then either use the sort you have or use lodash to shorten it by a couple lines.

File: test/helpers.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+exports.assertWorker = assertWorker;
+
+function liveWorkerIds() {
+  var workerIds = Object.keys(cluster.workers).filter(function(id) {
+    var worker = cluster.workers[id];
+    var isAlive = worker.process.signalCode == null &&
+      worker.process.exitCode == null;
+    return isAlive;
+  });
+  return workerIds.sort(function(a, b) {
```
Comment (rmg) [inline test/helpers.js]: This aspect of the behaviour isn't needed here except to turn up the volume on the code smell of reimplementing features in tests.

Perhaps if this `liveWorkerIds` is useful in multiple contexts it should be extracted to a separate module (node module, not npm module)?

File: test/restart-failed.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  var ended;
+
+  control.start({size: SIZE});
+
+  function size() {
+    return Object.keys(cluster.workers).length;
+  }
+
+  control.once('resize', function() {
+    if (ended) return;
```
Comment (rmg) [inline test/restart-failed.js]: Restart, but only once, and only if the test hasn't already ended by then? This handler has no assertions in it, is it for working around some sort of edge case?

File: test/restart-failed.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+var tap = require('tap');
+
+tap.test('good workers are not killed', function(t) {
+  var ended;
+
+  control.start({size: SIZE});
+
+  function size() {
+    return Object.keys(cluster.workers).length;
```
Comment (rmg) [inline test/restart-failed.js]: I would probably just inline this in the 1 place it is called, but as `_.size(cluster.workers)`;

File: test/test-resize-event.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+var master = require('../lib/master');
+var pickWorker = require('./helpers').pickWorker;
+var tap = require('tap');
+
+cluster.setupMaster({
+  exec: 'test/workers/null.js'
+});
+
+tap.test('should get resize event on return to configured size', function(t) {
+  function assertClusterResizesToConfiguredSizeAfter(somethingHappens, t) {
```
Comment (rmg) [inline test/test-resize-event.js]: This argument order feels wrong. Also, `doSomething` would be a better name I think.

File: test/test-resize-event.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  t.test('after worker destroy', function(t) {
+    assertClusterResizesToConfiguredSizeAfter(function(done) {
+      pickWorker()
+      .once('exit', done)
+      .destroy('SIGKILL'); // .destroy is other, better, name for .kill
+    }, t);
+  });
+
+  t.test('after worker signal', function(t) {
+    // this.timeout(5000); // exit is abnormal, and triggers throttling
```
Comment (rmg) [inline test/test-resize-event.js]: dead comment?

File: test/test-resize-from-1-to-0.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+var cluster = require('cluster');
+var master = require('../lib/master');
+var tap = require('tap');
+var workerCount = require('./helpers').liveWorkers;
+
+cluster.setupMaster({
+  exec: 'test/workers/null.js'
```
Comment (rmg) [inline test/test-resize-from-1-to-0.js]: Is this to get around the tap before cluster.isWorker problem?

File: test/test-resize-to-last-concurrent.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  exec: 'test/workers/null.js'
+});
+
+tap.test('to last of concurrent resizes', function(t) {
+  master.start({size: 10});
+  master.on('startWorker', function() {
+    if (workerCount() === 3) {
+      master.setSize(5);
+    }
+    if (workerCount() === 5) {
```
Comment (rmg) [inline test/test-resize-to-last-concurrent.js]: Is the lack of `else` because `master.setSize(5);` is expected to be synchronous and both of these conditions can actually be hit?

File: test/test-resize-while-forking.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  exec: 'test/workers/null.js'
+});
+
+tap.test('while workers are forked', function(t) {
+  master.start({size: 10});
+  master.on('startWorker', function() {
+    if (workerCount() === 1) {
+      cluster.fork();
+    }
+    if (workerCount() === 3) {
```
Comment (rmg) [inline test/test-resize-while-forking.js]: Similar to another test.. it almost feels like `workerCount()` is intentionally called multiple times. Are there intended/unintended side effects to doing so?

File: test/test-restart-all.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+cluster.setupMaster({
+  exec: 'test/workers/null.js'
+});
+
+tap.test('restart all workers', function(t) {
+  var SIZE = 5;
+  var DELAY = 200;
+
+  t.plan(4);
```
Comment (rmg) [inline test/test-restart-all.js]: This doesn't look like a test that requires a plan - is the flow less deterministic than it looks? If so, maybe a comment explaining how/why that is?

**Comment by rmg** [comment]
Couple trivial nits and comment/naming suggestions, but also a few questions raised by some of the test code. Could be addressed with minor code changes or comments.

File: test/_mystery-exit.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  };
+  dump.handles();
+  dump.requests();
+  process.on('exit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  process.on('beforeExit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  console.log('bye?');
```
Comment (sam-github) [inline test/_mystery-exit.js]: @bnoordhuis could you help me with this mystery, please? I am failing to understand why this script exits after logging "bye?". It has 3 servers listening, an immediate, a next tick, and a set timeout, and it exits, WITHOUT emitting the exit and beforeExit events, AFAICT. I'm not sure which is stranger, that the loop is dead, or that the events aren't emitted.

If I comment out the first tap test in the file, though, this second one runs to its expected completion. I'm not sure if I'm triggering a node bug, or if I'm just somehow damaging the node runtime (but how?).  I've even stepped through StartNodeInstance, and I can see the process emits, but nothing is printed. And uv definitely thinks its dead (according to uv__loop_alive).

Any ideas? Or even any ideas how to debug this? I'm at my wits end. For the purposes of the tests, I just split the tests up one per file, which is better test structure anyway, but its quite disconcerting to have node exiting like this - there is some kind of nasty bug somewhere, or I'm really misunderstanding something.

Output of above, FYI, on node 3.x (the handles active depends on node version, but all the versions I tested exit immediately despite the pending ticks, etc.).

```
sn/strong-cluster-control (fix-restart *%) % node test/_mystery-exit.js
TAP version 13
    # Subtest: from size 1 to 0
    ok 1 - worker is valid
    1..1
ok 1 - from size 1 to 0 # time=117.402ms

    # Subtest: from size 3 to 1
trying to stay alive
why do we die :-(
ActiveHandles:
0: process.stderr/type=tty/fd=2
1: process.stdout/type=tty/fd=1
2: timer:msecs=3000:count=1
ActiveRequests: []
bye?
```

File: lib/master.js (hunk 1 above)
Comment (sam-github) [inline lib/master.js]: done

File: test/restart-failed.js (hunk 3 above)
Comment (sam-github) [inline test/restart-failed.js]: no, resize event name is probably misleading, it would be better called "reached-configured-size", or something. In other words
1. start
2. resize event occurs once SIZE workers are started
3. call restart

The if ended isn't necessary, I'll remove, its a vestige of when it was `control.on('resize'`.

File: test/_mystery-exit.js (hunk 11 above)
Comment (rmg) [inline test/_mystery-exit.js]: Notice the assertion on line 21 is missing from your test output. I don't know how the first subtest is ending, but it's not from `master.stop(t.end);`.

File: test/test-resize-from-1-to-0.js (hunk 7 above)
Comment (sam-github) [inline test/test-resize-from-1-to-0.js]: Its the original code, all the refactored tests use the null worker, some of the new tests don't.

File: test/test-resize-event.js (hunk 5 above)
Comment (sam-github) [inline test/test-resize-event.js]: This is pretty cosmetic, its original code from before the refactor, I don't want to rewrite, just convert to tap.

File: test/test-resize-event.js (hunk 6 above)
Comment (sam-github) [inline test/test-resize-event.js]: removed

File: test/test-resize-to-last-concurrent.js (hunk 8 above)
Comment (sam-github) [inline test/test-resize-to-last-concurrent.js]: worker count is 3, or its 5, so no need for else

setSize is async, startWorker is emitted every time a worker is started, for each one. this is a series of set sizes, each happening when the last set size is partially complete

File: test/test-resize-while-forking.js (hunk 9 above)
Comment (sam-github) [inline test/test-resize-while-forking.js]: workerCount has no side effects, it just counts the current workers.

File: test/test-restart-all.js (hunk 10 above)
Comment (sam-github) [inline test/test-restart-all.js]: It needs a plan. It has two events, the test isn't over until they are both emitted. If you didn't have a plan, put the t.end() before the master.stop, and you deleted the code that emits the 'startRestart' event, then the test would pass, but without ever having seen that event, or done the deepEqual on its argument.

File: test/helpers.js (hunk 2 above)
Comment (sam-github) [inline test/helpers.js]: done

File: test/_mystery-exit.js (hunk 11 above)
Comment (sam-github) [inline test/_mystery-exit.js]: fascinating. you are right, the test is ending, but not because t.end was called. I think it ended because there was nothing left to do, and tap (somehow? wtf?) decided the test was over, or more likely, in an on-exit handler, just wrote out the the test was over... but then the second test was actually called. And it did hit the console.logs right after starting more async handles. so, hm, still baffling.

Comment (sam-github) [inline test/_mystery-exit.js]: so, fixing that does indeed cause both tests to run to completion. see _mystery-exit-2.js. I still don't get how I can start 4 kinds of async, and node still exits, without calling exit or reallyExit

Comment (rmg) [inline test/_mystery-exit.js]: It seems like the process is exiting but then allowing one more tick to be processed before the final exit(3) call. Perhaps this is an unintended side effect of the cluster module?

File: test/test-resize-to-last-concurrent.js (hunk 8 above)
Comment (rmg) [inline test/test-resize-to-last-concurrent.js]: My comment was more of a "what does this code make me think?" in case you wanted to address it. It's usually more a reflection of personal style differences and be ignored :-)

**Comment by rmg** [comment]
My re-review wasn't as thorough as my first review, but LGTM.

File: test/_mystery-exit.js (hunk 11 above)
Comment (sam-github) [inline test/_mystery-exit.js]: yes, that's what is happening, thanks for the suggestion.

tap is trapping the exit event, and in the exit event, running the next test. this doesn't go so well.

**Comment by sam-github** [comment]
@slnode test please

**Comment by rmg** [comment]
This appears to have broken strong-supervisor: http://ci.strongloop.com/job/strong-supervisor/1310/

File: lib/master.js (hunk 1 above)
Comment (sam-github) [inline lib/master.js]: It turns out that while js sorts numbers as if they were strings, lodash sorts strings as if they were lexical... and cluster IDs are strings, so lodash sorts as `["10", "9"]`... I should have investigated more thoroughly. PR to fix this is coming.
